### PR TITLE
Add tests for C# and VB NavigateTo

### PR DESCRIPTION
**Pull request by rchande**
cc @dotnet/roslyn-ide for review

### CODE REVIEW

**Comment by rchande** [comment]
Note that one scenario was removed from CSharpNavigateTo.xml and its replacement is tracked by https://github.com/dotnet/roslyn/issues/17531

**Comment by rchande** [comment]
retest windows_release_unit32_prtest  please

**Comment by rchande** [comment]
windows_release_unit32_prtest failed with some jenkins issue

File: src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs [hunk 1]
```diff
@@ -441,6 +441,15 @@ public void VerifyProjectReferencePresent(string projectName, string referencedP
...
             Assert.Contains(referencedProjectName, projectReferences);
+
+        }
+        protected void InvokeNavigateTo(string text)
+        {
+            ExecuteCommand(WellKnownCommandNames.Edit_GoToAll);
+            Editor.NavigateToSendKeys(text);
+            WaitForAsyncOperations(FeatureAttribute.NavigateTo);
+            Editor.NavigateToSendKeys("{ENTER}");
+            WaitForAsyncOperations(FeatureAttribute.NavigateTo);
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs]: Why the wait after the send keys?

File: src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs [hunk 2]
```diff
@@ -441,6 +441,15 @@ public void VerifyProjectReferencePresent(string projectName, string referencedP
         {
             var projectReferences = VisualStudio.Instance.SolutionExplorer.GetProjectReferences(projectName);
             Assert.Contains(referencedProjectName, projectReferences);
+
+        }
+        protected void InvokeNavigateTo(string text)
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs]: "InvokeNavigateToAndPressEnter"? I would have imagined this was only triggering it by the initial name.

Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs]: protected? Not public? What's the intended pattern here @tannergooding?

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+using Microsoft.VisualStudio.IntegrationTest.Utilities;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.Common;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.Input;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Roslyn.VisualStudio.IntegrationTests.CSharp
+{
+    [Collection(nameof(SharedIntegrationHostFixture))]
+    public class CSharpNavigateTo: AbstractEditorTest
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs]: Space before :.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        protected override string LanguageName => LanguageNames.CSharp;
+
+        public CSharpNavigateTo(VisualStudioInstanceFactory instanceFactory)
+            : base(instanceFactory, nameof(CSharpNavigateTo))
+        {
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.SignatureHelp)]
+        public void NavigateTo1()
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs]: Is there a NavigateTo2?

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+class FirstClass
+{
+    void FirstMethod() { }
+}");
+
+
+            AddFile("test2.cs", open: true, contents: @"
+");
+
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs]: Please no waiting on workspaces directly. Please move it to the helper that needs the wait, and justify it's need to be there.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            InvokeNavigateTo("FirstClass");
+            Assert.Equal("FirstClass", Editor.GetSelectedText());
+            InvokeNavigateTo("FirstMethod");
+            Assert.Equal("FirstMethod", Editor.GetSelectedText());
+
+            VisualStudio.Instance.SolutionExplorer.AddProject("VBProject", WellKnownProjectTemplates.ClassLibrary, LanguageNames.VisualBasic);
+            VisualStudio.Instance.SolutionExplorer.AddFile("VBProject", "vbfile.vb", open: true);
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs]: Ditto: if there's a need, let's move it to the underlying helper. I suspect you can just delete this, and this is remnants back when the workspace model was different.

File: src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+using Microsoft.VisualStudio.IntegrationTest.Utilities;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.Common;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.Input;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Roslyn.VisualStudio.IntegrationTests.CSharp
+{
+    [Collection(nameof(SharedIntegrationHostFixture))]
+    public class BasicNavigateTo: AbstractEditorTest
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs]: Space before :.

File: src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.IntegrationTest.Utilities;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.Common;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.Input;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Roslyn.VisualStudio.IntegrationTests.CSharp
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs]: Wrong namespace?

File: src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+Class FirstClass
+    Sub FirstMethod()
+    End Sub
+End Class");
+
+
+            AddFile("test2.cs", open: true, contents: @"
+");
+
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs]: Ditto.

File: src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+            InvokeNavigateTo("FirstClass");
+            Assert.Equal("FirstClass", Editor.GetSelectedText());
+            InvokeNavigateTo("FirstMethod");
+            Assert.Equal("FirstMethod", Editor.GetSelectedText());
+
+            VisualStudio.Instance.SolutionExplorer.AddProject("CSProject", WellKnownProjectTemplates.ClassLibrary, LanguageNames.CSharp);
+            VisualStudio.Instance.SolutionExplorer.AddFile("CSProject", "csfile.cs", open: true);
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs]: Ditto.

File: src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            AddFile("test2.cs", open: true, contents: @"
+");
+
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
+            InvokeNavigateTo("FirstClass");
+            Assert.Equal("FirstClass", Editor.GetSelectedText());
+            InvokeNavigateTo("FirstMethod");
+            Assert.Equal("FirstMethod", Editor.GetSelectedText());
+
+            VisualStudio.Instance.SolutionExplorer.AddProject("CSProject", WellKnownProjectTemplates.ClassLibrary, LanguageNames.CSharp);
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs]: Why are we adding the _other_ language in the test? Ditto for the other one.

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs [hunk 12]
```diff
@@ -524,6 +532,18 @@ public void DialogSendKeys(string dialogAutomationName, string keys)
             SendKeys.SendWait(keys);
         }
 
+        public void NavigateToSendKeys(string keys)
+        {
+            var dialogAutomationElement = FindNavigateTo();
+            if (dialogAutomationElement == null)
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs]: Can this ever happen? Pretty sure the underlying Finds will go forever.

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs [hunk 13]
```diff
@@ -524,6 +532,18 @@ public void DialogSendKeys(string dialogAutomationName, string keys)
             SendKeys.SendWait(keys);
         }
 
+        public void NavigateToSendKeys(string keys)
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs]: "SendKeysToNavigateTo"?

File: src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs (hunk 2 above)
Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs]: Same here.

Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs]: The expectation is that the test class derive from `AbstractEditorTest` thus providing access.

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs [hunk 14]
```diff
@@ -548,6 +568,13 @@ private static AutomationElement FindDialogWorker(string dialogAutomationName)
             return vsAutomationElement.FindFirst(TreeScope.Descendants, elementCondition);
         }
 
+        private static AutomationElement FindNavigateTo()
+        {
+            var vsAutomationElement = AutomationElement.FromHandle(new IntPtr(GetDTE().MainWindow.HWnd));
+            Condition elementCondition = new PropertyCondition(AutomationElement.AutomationIdProperty, "PART_SearchBox");
+            return vsAutomationElement.FindFirst(TreeScope.Descendants, elementCondition);
+        }
```
Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs]: Use `AutomationElementExtensions`.

File: src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs (hunk 11 above)
Comment (rchande) [inline src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicNavigateTo.cs]: That's what the old integration tests did. I guess they wanted to make sure NavigateTo worked "cross language"?

**Comment by rchande** [comment]
retest this please

**Comment by heejaechang** [comment]
retest this please

**Comment by rchande** [comment]
@tmeschter @jasonmalinowski I believe I've incorporated all of your feedback if you'd like to take another look. Thanks!

**Comment by rchande** [comment]
Looks like the failures are in BasicWinForms and unrelated.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+        public CSharpNavigateTo(VisualStudioInstanceFactory instanceFactory)
+            : base(instanceFactory, nameof(CSharpNavigateTo))
+        {
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.SignatureHelp)]
+        public void NavigateTo()
+        {
+            AddFile("test1.cs", open:false, contents: @"
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs]: Space after `open:`.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+");
+
+            InvokeNavigateToAndPressEnter("FirstClass");
+            Editor.WaitForActiveView("test1.cs");
+            Assert.Equal("FirstClass", Editor.GetSelectedText());
+            InvokeNavigateToAndPressEnter("FirstMethod");
+            Editor.WaitForActiveView("test1.cs");
+            Assert.Equal("FirstMethod", Editor.GetSelectedText());
+
+            VisualStudio.Instance.SolutionExplorer.AddProject("VBProject", WellKnownProjectTemplates.ClassLibrary, LanguageNames.VisualBasic);
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs]: I don't get what this is testing...? Can we add a comment? Is this supposed to be finding things in the VB projects too?

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs [hunk 17]
```diff
@@ -524,6 +537,18 @@ public void DialogSendKeys(string dialogAutomationName, string keys)
             SendKeys.SendWait(keys);
         }
 
+        public void SendKeysToNavigateTo(string keys)
+        {
+            var dialogAutomationElement = FindNavigateTo();
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs]: Does FindNavigateTo spin forever? If so, why the null check later?

File: src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/Editor_OutOfProc.cs [hunk 18]
```diff
@@ -141,5 +144,22 @@ public void PressDialogButton(string dialogAutomationName, string buttonAutomati
...
             => _inProc.DialogSendKeys(dialogAutomationName, keys);
+
+        public void NavigateToSendKeys(string keys)
+            => _inProc.SendKeysToNavigateTo(keys);
+
+        public void WaitForActiveView(string viewName)
+        {
+            while (true)
+            {
+                System.Threading.Thread.Sleep(100);
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/Editor_OutOfProc.cs]: There's some generic helper for "retry until" pattern that we should reuse. That way we can put timeout logic in one place.

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs (hunk 17 above)
Comment (rchande) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs]: It does not spin.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs (hunk 16 above)
Comment (rchande) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpNavigateTo.cs]: Done.

**Comment by rchande** [comment]
retest windows_debug_vs-integration_prtest

**Comment by rchande** [comment]
retest windows_debug_vs-integration_prtest please

**Pull request merged**